### PR TITLE
[Owner Review] Generated and built APIs for Digital Pattern

### PR DESCRIPTION
**Pull request by ShantanuShinde**
### What does this Pull Request accomplish?

Added metadata for Digital pattern from nimi-python and hapigen, generated the APIs and built them.
111/137 APIs generating so far

### Why should this Pull Request be merged?

The APIs for Digital Pattern are generating and building

### What testing has been done?

Build successful

### CODE REVIEW

File: generated/nidigital/nidigital.proto [hunk 1]
```diff
@@ -0,0 +1,1677 @@
...
+
+enum ExportSignal {
+  EXPORT_SIGNAL_UNSPECIFIED = 0;
+  EXPORT_SIGNAL_NIDIGITAL_VAL_START_TRIGGER = 2000;
+  EXPORT_SIGNAL_NIDIGITAL_VAL_CONDITIONAL_JUMP_TRIGGER = 2001;
+  EXPORT_SIGNAL_NIDIGITAL_VAL_PATTERN_OPCODE_EVENT = 2002;
+  EXPORT_SIGNAL_NIDIGITAL_VAL_REF_CLOCK = 2003;
+}
+
+enum HistoryRAMCyclesToAcquire {
```
Comment (doshirohan) [inline generated/nidigital/nidigital.proto]: According to Teams discussion, the acronym should be camel cased `HistoryRamCyclesToAcquire`. Make this changes to all other acronyms in the `enums.py` as well and document in CHANGES.md file.

File: source/codegen/metadata/nidigital/functions_addon.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
+# These dictionaries are merged with the extracted function metadata at build time.
```
Comment (doshirohan) [inline source/codegen/metadata/nidigital/functions_addon.py:1]: I believe our auto-generation scripts do not use `_addon` files and I don't believe any of our current drivers had significant info in addon files. @kruthi-NI @Akshata440 do we need to take a look at this additional metadata and what it means for our codegen scripts or for creating additional tasks to implement these methods?

File: source/server/core_server.cpp [hunk 3]
```diff
@@ -1,9 +1,12 @@
+
```
Comment (doshirohan) [inline source/server/core_server.cpp]: Remove the empty line.

File: source/server/core_server.cpp [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
 #include <niscope/niscope_library.h>
 #include <niscope/niscope_service.h>
 #include <niswitch/niswitch_library.h>
 #include <niswitch/niswitch_service.h>
 #include <nisync/nisync_library.h>
 #include <nisync/nisync_service.h>
+#include <nidigital/nidigital_library.h>
```
Comment (doshirohan) [inline source/server/core_server.cpp]: Can you add these two lines towards beginning of file to keep the include statements sorted in alphabetical order.

**Review by doshirohan — COMMENTED** [review]
I want to wait on Kruthi/Akshata's comment on functions_addon.py before approving the PR. I will approve based on their response.

File: generated/nidigital/nidigital_library_interface.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
+//---------------------------------------------------------------------
+// This file is automatically generated. All manual edits will be lost.
+//---------------------------------------------------------------------
+// Library wrapper for implementing interactions with NI-Digital Pattern Driver
+//---------------------------------------------------------------------
+#ifndef NIDIGITAL_GRPC_LIBRARY_WRAPPER_H
+#define NIDIGITAL_GRPC_LIBRARY_WRAPPER_H
+
+#include <grpcpp/grpcpp.h>
+#include <niDigital.h>
```
Comment (doshirohan) [inline generated/nidigital/nidigital_library_interface.h:10]: You need to copy copy `niDigital.h` file in imports folder and submit that as part of this PR.

File: generated/nidigital/nidigital.proto (hunk 1 above)
Comment (ShantanuShinde) [reply generated/nidigital/nidigital.proto]: fixed that

File: source/server/core_server.cpp (hunk 4 above)
Comment (ShantanuShinde) [reply source/server/core_server.cpp]: fixed this

File: generated/nidigital/nidigital_library_interface.h (hunk 5 above)
Comment (ShantanuShinde) [reply generated/nidigital/nidigital_library_interface.h:10]: the import folder got left out of commit by mistake. fixed that

**Review by doshirohan — APPROVED** [review]

File: source/codegen/metadata/nidigital/functions_addon.py (hunk 2 above)
Comment (ShantanuShinde) [reply source/codegen/metadata/nidigital/functions_addon.py:1]: A separate task will be created regarding the  _addon data

File: source/codegen/templates/service_helpers.mako [hunk 6]
```diff
@@ -187,6 +187,8 @@ one_of_case_prefix = f'{namespace_prefix}{function_name}Request::{PascalFieldNam
 % elif c_type == 'ViSession':
       auto ${parameter_name}_grpc_session = ${request_snippet};
       ${c_type} ${parameter_name} = session_repository_->access_session(${parameter_name}_grpc_session.id(), ${parameter_name}_grpc_session.name());\
+% elif c_type == 'ViUInt32[]' or c_type == 'ViInt32[]':
```
Comment (jonlugoatx) [inline source/codegen/templates/service_helpers.mako]: I don't think a `const_cast` of a `reinterpret_cast` is the right implementation. I thought we already supported `ViInt32[]` and `ViUInt32[]`. The `reinterpret_cast` can easily hide bugs. What's the error you're seeing with the `const_cast`?

**Review by jonlugoatx — CHANGES_REQUESTED** [review]

File: source/codegen/templates/service_helpers.mako (hunk 6 above)
Comment (reckenro) [reply source/codegen/templates/service_helpers.mako]: Also, if you do end up making code generation changes I'd prefer them in a separate PR so we can have a separate view into

1. Metadata changes for already supported code generation patterns and resulting generated code
2. Code generation changes with only metadata updates that consume the new codegen support.

File: source/codegen/metadata/nidigital/functions.py [hunk 7]
```diff
@@ -0,0 +1,3082 @@
...
+            {
+                "direction": "in",
+                "name": "dataBufferSize",
+                "type": "ViInt32"
+            },
+            {
+                "direction": "out",
+                "name": "data",
+                "size": {
+                    "mechanism": "ivi-dance-with-a-twist",
```
Comment (reckenro) [inline source/codegen/metadata/nidigital/functions.py]: [Note] You probably have this captured but just wanted to raise awareness that any method right now with `ivi-dance-with-a-twist` mechanism will generate a simple service method returning a `::grpc::UNIMPLEMENTED` status.

**Review by reckenro — CHANGES_REQUESTED** [review]
Curious to see the resolution of the comments in `service_helpers.mako` and want to take another pass after.

**Review by SparkingSpork — CHANGES_REQUESTED** [review]
I don't see any issues that haven't been already called out by others but would like to be reset when those issues are resolved.

File: source/codegen/templates/service_helpers.mako (hunk 6 above)
Comment (ShantanuShinde) [reply source/codegen/templates/service_helpers.mako]: @jonlugoatx I am getting this error when using `const_cast`
image

Comment (ShantanuShinde) [reply source/codegen/templates/service_helpers.mako]: I have removed the changes and the APIs requiring it. they will be added in a separate PR

File: source/codegen/metadata/nidigital/functions.py (hunk 7 above)
Comment (ShantanuShinde) [reply source/codegen/metadata/nidigital/functions.py]: I removed those APIs. they will be added in a separate PR

**Review by SparkingSpork — APPROVED** [review]

**Review by reckenro — APPROVED** [review]

File: source/codegen/templates/service_helpers.mako (hunk 6 above)
Comment (jonlugoatx) [reply source/codegen/templates/service_helpers.mako]: The important part of the error is what comes after the "with". A `const google::protobuf::uint32*` which is really a `const uint32_t*` should be convertible to a `ViUInt32*` using a `const_cast`. Are we sure the proto file is using `repeated uint32` for `ViUInt32[]` types?

**Review by jonlugoatx — APPROVED** [review]

File: source/codegen/templates/service_helpers.mako (hunk 6 above)
Comment (ShantanuShinde) [reply source/codegen/templates/service_helpers.mako]: I will give that a look and follow up in the PR adding those APIs

**Pull request merged**